tiledb/sm/query/query_remote_buffer_storage.cc

```diff
@@ -0,0 +1,277 @@
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This file defines class query_remote_buffer_storage.
+ */
+
+#include "tiledb/sm/query/query_remote_buffer_storage.h"
+#include "tiledb/sm/array/array.h"
+#include "tiledb/sm/query/query.h"
+
+namespace tiledb::sm {
+class QueryRemoteBufferStorageException : public StatusException {
+ public:
+  explicit QueryRemoteBufferStorageException(const std::string& msg)
+      : StatusException("QueryRemoteBufferStorage", msg) {
+  }
+};
+
+/* ***************************** */
+/*         BufferCache           */
+/* ***************************** */
+
+void BufferCache::adjust_offsets(uint64_t cached_bytes) {
+  if (buffer_.size() < cached_bytes) {
+    throw QueryRemoteBufferStorageException(
+        "Attempt to correct offsets outside of fixed buffer bounds.");
+  }
+  // Adjust any appended offsets to be relative to already cached data.
+  for (uint64_t pos = buffer_.size() - cached_bytes; pos < buffer_.size();
+       pos += cell_size_) {
+    *(uint64_t*)buffer_.data(pos) += buffer_var_.size();
+  }
+}
+
+/* ***************************** */
+/*   QueryRemoteBufferStorage    */
+/* ***************************** */
+
+QueryRemoteBufferStorage::QueryRemoteBufferStorage(Query* query)
+    : query_(query) {
+  if (!query_->array()->is_remote()) {
+    return;
+  }
+  cell_num_per_tile_ = query_->array_schema().dense() ?
+                           query_->array_schema().domain().cell_num_per_tile() :
+                           query_->array_schema().capacity();
+
+  // Only allocate cache buffers for remote global order writes.
+  // For other types or layouts the keys will exist with empty buffers.
+  if (query_->type() == QueryType::READ ||
+      query_->layout() != Layout::GLOBAL_ORDER) {
+    cell_num_per_tile_ = 0;
+  }
+
+  for (const auto& attr : query_->array_schema().attributes()) {
+    std::string name = attr->name();
+    bool is_var = query_->array_schema().var_size(name);
+    bool is_nullable = query_->array_schema().is_nullable(name);
+    uint64_t data_size = datatype_size(query_->array_schema().type(name));
+    uint64_t cell_size = is_var ? constants::cell_var_offset_size : data_size;
+
+    caches_[name].cell_size_ = cell_size;
+    // Construct and move a preallocated Buffer to initialize cache.
+    caches_[name].buffer_ = {cell_num_per_tile_ * cell_size};
+    if (is_nullable) {
+      caches_[name].buffer_validity_ = {cell_num_per_tile_};
+    }
+  }
+
+  // Initialize coordinate buffers for sparse arrays.
+  if (!query_->is_dense()) {
+    for (const auto& name : query_->array_schema().dim_names()) {
+      uint64_t cell_size = datatype_size(query_->array_schema().type(name));
```

Comment:
This would not have worked for var-sized dimensions. Good catch! With the other changes requested here, I was able to combine this with the loop above, which would work for var-size dimensions.